File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left = 0 
        right = len(nums)-1
        mid = 0
        while left < right:
            mid = (right + left) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            # elif nums[left] > nums[right] and nums[right] > nums[mid]:
            #     if  nums[mid+1] > nums[mid] and nums[mid-1] > nums[mid]:
            #         return nums[mid]
            #     else:
            #         right = mid - 1
            else:
                right = mid
        return nums[left]
        
        

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.cpp
class Solution {
public:
// A..B...C...Z note you need a window that allows for k rotations to happen
// if it goes beyond, then you need to change window 
// left ==> maxFreq ==> left to flip ==> right 
    int characterReplacement(string s, int k) {
        vector<int> count(26,0);
        int left = 0;
        int maxFreq = 0;
        int window = 0;
        for(int r = 0; r < s.size(); r++){
            count[s[r] - 'A']++;
            maxFreq = max(maxFreq, count[s[r]-'A']);
            while((r - left + 1) - maxFreq > k) { // slide window if the window reached max
                count[s[left] -'A']--;
                left++;
            }
            window = max(window, r - left + 1);
        }
        return window;
    }
};
